File: backend/lib/print/letter_size_uniformity.m
function results = compute_letter_size_uniformity(image_path, debug)
    % Set default for debug if not provided
    if nargin < 2
        debug = false;
    end

    % Read the image
    img = imread(image_path);
    if isempty(img)
        disp(['Error: Could not read image at ', image_path]);
        results = struct();
        return;
    end

    % Convert to grayscale if image is RGB
    if size(img, 3) == 3
        gray = rgb2gray(img);
    else
        gray = img;
    end

    % Perform analysis on original grayscale for intensity measurements
    % We'll use this to analyze pen pressure through pixel intensity

    % Apply binary thresholding and invert for component analysis
    threshold = 127 / 255;
    binary = imbinarize(gray, threshold);
    binary = ~binary;  % Inversion to match cv2.THRESH_BINARY_INV

    % Find connected components (letters/characters)
    cc = bwconncomp(binary);
    stats = regionprops(cc, 'BoundingBox', 'Area', 'Centroid', 'PixelIdxList');
    
    % Filter out small noise components
    min_area = 20;  % Minimum area threshold
    valid_indices = [];
    for i = 1:length(stats)
        if stats(i).Area > min_area
            valid_indices = [valid_indices, i];
        end
    end
    stats = stats(valid_indices);
    
    % Extract heights and widths of bounding boxes
    heights = zeros(1, length(stats));
    widths = zeros(1, length(stats));
    aspect_ratios = zeros(1, length(stats));
    
    % For pen pressure analysis
    mean_intensities = zeros(1, length(stats));
    intensity_variations = zeros(1, length(stats));
    stroke_widths = zeros(1, length(stats));
    
    for i = 1:length(stats)
        % Get bounding box dimensions
        bbox = stats(i).BoundingBox;
        widths(i) = bbox(3);
        heights(i) = bbox(4);
        aspect_ratios(i) = widths(i) / heights(i);
        
        % Get pixel indices for this component
        pixelIdx = stats(i).PixelIdxList;
        
        % Extract intensity values for this letter
        letter_intensity = 255 - double(gray(pixelIdx));  % Invert values so higher = darker
        
        % Calculate mean intensity as a measure of pen pressure
        mean_intensities(i) = mean(letter_intensity);
        
        % Calculate variation in intensity within the letter
        intensity_variations(i) = std(letter_intensity);
        
        % Estimate stroke width using distance transform
        % Create a mask for just this letter
        letter_mask = false(size(binary));
        letter_mask(pixelIdx) = true;
        
        % Distance transform to find the distance to the nearest background pixel
        dist_transform = bwdist(~letter_mask);
        
        % The maximum distance gives half the maximum stroke width
        max_dist = max(dist_transform(pixelIdx));
        stroke_widths(i) = 2 * max_dist;  % Double to get full width
    end
    
    % Check if any letters were found
    if isempty(heights)
        results = struct('height_uniformity', 0, ...
                        'width_uniformity', 0, ...
                        'aspect_ratio_uniformity', 0, ...
                        'pen_pressure_uniformity', 0, ...
                        'stroke_width_uniformity', 0, ...
                        'avg_pen_pressure', 0, ...
                        'avg_stroke_width', 0, ...
                        'letter_count', 0);
        return;
    end
    
    % Calculate uniformity as 1 - coefficient of variation
    % Lower variation means higher uniformity
    height_cv = std(heights) / mean(heights);
    width_cv = std(widths) / mean(widths);
    aspect_ratio_cv = std(aspect_ratios) / mean(aspect_ratios);
    
    % Pen pressure related metrics
    pressure_cv = std(mean_intensities) / mean(mean_intensities);
    stroke_width_cv = std(stroke_widths) / mean(stroke_widths);
    
    height_uniformity = max(0, 1 - height_cv);
    width_uniformity = max(0, 1 - width_cv);
    aspect_ratio_uniformity = max(0, 1 - aspect_ratio_cv);
    pen_pressure_uniformity = max(0, 1 - pressure_cv);
    stroke_width_uniformity = max(0, 1 - stroke_width_cv);
    
    % Pack results into a structure
    results = struct(...
        'height_uniformity', height_uniformity, ...
        'width_uniformity', width_uniformity, ...
        'aspect_ratio_uniformity', aspect_ratio_uniformity, ...
        'pen_pressure_uniformity', pen_pressure_uniformity, ...
        'stroke_width_uniformity', stroke_width_uniformity, ...
        'avg_pen_pressure', mean(mean_intensities) / 255, ...
        'avg_stroke_width', mean(stroke_widths), ...
        'letter_count', length(stats) ...
    );
    
    % Debug visualization if requested
    if debug
        figure('Name', 'Letter Size and Pen Pressure Analysis', 'NumberTitle', 'off');
        
        % Original Image
        subplot(2,3,1);
        imshow(img);
        title('Original Image');
        
        % Binary Image with Bounding Boxes
        subplot(2,3,2);
        imshow(binary);
        hold on;
        for i = 1:length(stats)
            bbox = stats(i).BoundingBox;
            rectangle('Position', bbox, 'EdgeColor', 'r', 'LineWidth', 1);
            % Add letter index for reference
            text(stats(i).Centroid(1), stats(i).Centroid(2), num2str(i), ...
                'Color', 'g', 'FontSize', 8, 'FontWeight', 'bold');
        end
        hold off;
        title('Letter Detection');
        
        % Height Distribution
        subplot(2,3,3);
        histogram(heights, min(20, ceil(length(heights)/2)));
        title(sprintf('Height Distribution\nUniformity: %.3f', height_uniformity));
        
        % Width Distribution
        subplot(2,3,4);
        histogram(widths, min(20, ceil(length(widths)/2)));
        title(sprintf('Width Distribution\nUniformity: %.3f', width_uniformity));
        
        % Pen Pressure Distribution
        subplot(2,3,5);
        histogram(mean_intensities, min(20, ceil(length(mean_intensities)/2)));
        title(sprintf('Pen Pressure Distribution\nUniformity: %.3f', pen_pressure_uniformity));
        
        % Stroke Width Distribution
        subplot(2,3,6);
        histogram(stroke_widths, min(20, ceil(length(stroke_widths)/2)));
        title(sprintf('Stroke Width Distribution\nUniformity: %.3f', stroke_width_uniformity));
        
        % Print debug information in the command window
        fprintf('Letter count: %d\n', length(stats));
        fprintf('Height uniformity: %.3f\n', height_uniformity);
        fprintf('Width uniformity: %.3f\n', width_uniformity);
        fprintf('Aspect ratio uniformity: %.3f\n', aspect_ratio_uniformity);
        fprintf('Pen pressure uniformity: %.3f\n', pen_pressure_uniformity);
        fprintf('Stroke width uniformity: %.3f\n', stroke_width_uniformity);
        fprintf('Average pen pressure (0-1): %.3f\n', results.avg_pen_pressure);
        fprintf('Average stroke width: %.3f\n', results.avg_stroke_width);
    end
end

% === Main Script ===
% Replace with your actual image file path.
image_path = '/Users/jameswong/PycharmProjects/NoteMercy_Extension/backend/atest/1.png';
uniformity_results = compute_letter_size_uniformity(image_path, true);
disp(uniformity_results);

File: backend/lib_py/print/letter_size_uniformity.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
from skimage import measure
from scipy.ndimage import distance_transform_edt
import base64
from io import BytesIO

class LetterUniformityAnalyzer:
    def __init__(self, image_input, is_base64=True):
        """
        Initializes the LetterUniformityAnalyzer with either a base64 encoded image or image path.

        Parameters:
            image_input (str): Either base64 encoded image string or image file path
            is_base64 (bool): If True, image_input is treated as base64 string, else as file path
        """
        if is_base64:
            # Decode base64 image
            img_data = base64.b64decode(image_input)
            nparr = np.frombuffer(img_data, np.uint8)
            self.img = cv2.imdecode(nparr, cv2.IMREAD_COLOR)
            if self.img is None:
                raise ValueError("Error: Could not decode base64 image")
        else:
            # Read image from file path
            self.img = cv2.imread(image_input)
            if self.img is None:
                raise ValueError(f"Error: Could not read image at {image_input}")

        # Convert to grayscale if needed
        if len(self.img.shape) == 3 and self.img.shape[2] == 3:
            self.gray_img = cv2.cvtColor(self.img, cv2.COLOR_BGR2GRAY)
        else:
            self.gray_img = self.img.copy()

    def analyze(self, debug=False):
        """
        Analyzes the image to determine letter size uniformity by measuring
        height, width, aspect ratio, pen pressure, and stroke width.

        Parameters:
            debug (bool): If True, generates visualization plots.

        Returns:
            dict: Metrics and visualization graphs if debug=True.
        """
        # -----------------------------
        # 1) Apply Otsu's thresholding
        # -----------------------------
        # Using Otsu automatically determines the best threshold for binarization
        # in images with bimodal histograms.
        blur = cv2.GaussianBlur(self.gray_img, (3, 3), 0)
        # Otsu’s threshold returns the threshold value as 'thresh_val'
        thresh_val, binary = cv2.threshold(
            blur, 0, 255, cv2.THRESH_BINARY_INV + cv2.THRESH_OTSU
        )

        # ---------------------------------------
        # 2) Morphological operations (optional)
        # ---------------------------------------
        # A small opening removes noise. Adjust kernel size as needed.
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (2, 2))
        binary = cv2.morphologyEx(binary, cv2.MORPH_OPEN, kernel, iterations=1)

        # Optionally, a small closing can unify broken strokes
        # binary = cv2.morphologyEx(binary, cv2.MORPH_CLOSE, kernel, iterations=1)

        # ---------------------------------------------------
        # 3) Label connected components (letters/characters)
        # ---------------------------------------------------
        label_img = measure.label(binary, connectivity=2)
        regions = measure.regionprops(label_img, intensity_image=self.gray_img)

        # --------------------------------------------------------------------------------
        # 4) Filter out small noise components using a dynamic minimum area threshold
        # --------------------------------------------------------------------------------
        # For example, we can compute a robust estimate (like median area) among all
        # connected components, then set some fraction of that as the threshold.
        all_areas = [r.area for r in regions]
        median_area = np.median(all_areas) if all_areas else 0
        # As a starting point, we set min_area to 0.2 * median_area, but at least 20.
        min_area = max(20, 0.2 * median_area)

        valid_regions = [region for region in regions if region.area >= min_area]

        # If no regions were found, return zeros
        if len(valid_regions) == 0:
            metrics = {
                'height_uniformity': 0,
                'width_uniformity': 0,
                'aspect_ratio_uniformity': 0,
                'pen_pressure_uniformity': 0,
                'stroke_width_uniformity': 0,
                'avg_pen_pressure': 0,
                'avg_stroke_width': 0,
                'letter_count': 0
            }
            return {'metrics': metrics, 'graphs': []}

        # ---------------------------------------------------------------------------
        # 5) Compute metrics (height, width, aspect ratio, stroke width, pen pressure)
        # ---------------------------------------------------------------------------
        heights = []
        widths = []
        aspect_ratios = []
        mean_intensities = []
        stroke_widths = []
        bboxes = []
        centroids = []

        for i, region in enumerate(valid_regions):
            # Get bounding box (min_row, min_col, max_row, max_col)
            min_row, min_col, max_row, max_col = region.bbox
            width = max_col - min_col
            height = max_row - min_row
            aspect_ratio = width / height if height != 0 else 0

            heights.append(height)
            widths.append(width)
            aspect_ratios.append(aspect_ratio)
            bboxes.append((min_col, min_row, width, height))
            centroids.append((region.centroid[1], region.centroid[0]))

            # Extract intensity values for this letter (invert so higher = darker)
            coords = region.coords
            letter_intensity = 255 - self.gray_img[coords[:, 0], coords[:, 1]].astype(np.float64)
            mean_intensity = np.mean(letter_intensity)
            mean_intensities.append(mean_intensity)

            # Estimate stroke width using distance transform
            mask = np.zeros(binary.shape, dtype=np.uint8)
            mask[coords[:, 0], coords[:, 1]] = 1
            dist_transform = distance_transform_edt(mask)
            max_dist = np.max(dist_transform[mask == 1])
            stroke_width = 2 * max_dist
            stroke_widths.append(stroke_width)

        # ---------------------------------------------------------------------------
        # 6) Compute uniformities using coefficient of variation (CV = std / mean)
        # ---------------------------------------------------------------------------
        def uniformity_score(values):
            if len(values) == 0 or np.mean(values) == 0:
                return 0
            cv = np.std(values) / np.mean(values)
            return max(0, 1 - cv)  # ensure we don't return negative

        height_uniformity = uniformity_score(heights)
        width_uniformity = uniformity_score(widths)
        aspect_ratio_uniformity = uniformity_score(aspect_ratios)
        pen_pressure_uniformity = uniformity_score(mean_intensities)
        stroke_width_uniformity = uniformity_score(stroke_widths)

        metrics = {
            'height_uniformity': height_uniformity,
            'width_uniformity': width_uniformity,
            'aspect_ratio_uniformity': aspect_ratio_uniformity,
            'pen_pressure_uniformity': pen_pressure_uniformity,
            'stroke_width_uniformity': stroke_width_uniformity,
            'avg_pen_pressure': np.mean(mean_intensities) / 255,
            'avg_stroke_width': np.mean(stroke_widths),
            'letter_count': len(valid_regions)
        }

        result = {
            'metrics': metrics,
            'graphs': []
        }

        # -----------------------------------------------------
        # 7) Optional Debug Visualization
        # -----------------------------------------------------
        if debug:
            plt.figure("Letter Size and Pen Pressure Analysis", figsize=(12, 8))

            # Original Image
            plt.subplot(2, 3, 1)
            plt.imshow(cv2.cvtColor(self.img, cv2.COLOR_BGR2RGB))
            plt.title("Original Image")
            plt.axis('off')

            # Binary Image with Bounding Boxes
            plt.subplot(2, 3, 2)
            plt.imshow(binary, cmap='gray')
            ax = plt.gca()
            for idx, (bbox, centroid) in enumerate(zip(bboxes, centroids)):
                rect = plt.Rectangle((bbox[0], bbox[1]), bbox[2], bbox[3],
                                     edgecolor='r', facecolor='none', linewidth=1)
                ax.add_patch(rect)
                plt.text(centroid[0], centroid[1], str(idx + 1),
                         color='g', fontsize=8, fontweight='bold')
            plt.title("Letter Detection")
            plt.axis('off')

            # Height Distribution
            plt.subplot(2, 3, 3)
            bins = min(20, int(np.ceil(len(heights) / 2)) or 1)
            plt.hist(heights, bins=bins)
            plt.title(f"Height Distribution\nUniformity: {height_uniformity:.3f}")

            # Width Distribution
            plt.subplot(2, 3, 4)
            bins = min(20, int(np.ceil(len(widths) / 2)) or 1)
            plt.hist(widths, bins=bins)
            plt.title(f"Width Distribution\nUniformity: {width_uniformity:.3f}")

            # Pen Pressure Distribution
            plt.subplot(2, 3, 5)
            bins = min(20, int(np.ceil(len(mean_intensities) / 2)) or 1)
            plt.hist(mean_intensities, bins=bins)
            plt.title(f"Pen Pressure Distribution\nUniformity: {pen_pressure_uniformity:.3f}")

            # Stroke Width Distribution
            plt.subplot(2, 3, 6)
            bins = min(20, int(np.ceil(len(stroke_widths) / 2)) or 1)
            plt.hist(stroke_widths, bins=bins)
            plt.title(f"Stroke Width Distribution\nUniformity: {stroke_width_uniformity:.3f}")

            plt.tight_layout()

            # Convert plot to base64
            buf = BytesIO()
            plt.savefig(buf, format='png', bbox_inches='tight')
            buf.seek(0)
            plot_base64 = base64.b64encode(buf.getvalue()).decode('utf-8')
            plt.close()

            result['graphs'].append(plot_base64)

        return result

# === Example usage ===
if __name__ == "__main__":
    # Example with file path
    image_path = '../../atest/print.jpg'
    analyzer = LetterUniformityAnalyzer(image_path, is_base64=False)
    results = analyzer.analyze(debug=True)
    print(results['metrics'])
